### ISSUE TITLE
Recover with non-retried exceptions works with 1.2.2.RELEASE but not 1.2.3 or 1.2.4

### ISSUE THREAD

**Issue by adityaparikh91087** [spring-projects/spring-retry#169]
```
 @Override
  @Retryable(value = {RetriableException.class}, maxAttempts = 5)
  public Event process(Command command) throws NonRetriableException, AnotherNonRetriableException{
    try {
      return update(command);
    } catch (AnotherNonRetriableExceptione) {
      return updateFailed(command, e);
    }
  }

 @Recover
  public Event recover(Exception e, Command command) throws Exception {
    if(e instanceof RetriableException) {
      return updateFailed(command, e);
    }
    else {
      throw e;
    }
  }


```
where update() can throw RetriableException and NonRetriableException RuntimeExceptions
Works with version 1.2.2.RELEASE. Breaks when upgraded to  <spring-boot.version>2.1.3.RELEASE</spring-boot.version>

that brought in 1.2.4.RELEASE with ArrayIndexOutOfBoundsException

**Comment by adityaparikh91087** [comment]
I get the following stacktrace: 
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: -1
	at org.springframework.retry.annotation.RecoverAnnotationRecoveryHandler.compareParameters(RecoverAnnotationRecoveryHandler.java:124)
	at org.springframework.retry.annotation.RecoverAnnotationRecoveryHandler.findClosestMatch(RecoverAnnotationRecoveryHandler.java:95)
	at org.springframework.retry.annotation.RecoverAnnotationRecoveryHandler.recover(RecoverAnnotationRecoveryHandler.java:59)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor$ItemRecovererCallback.recover(RetryOperationsInterceptor.java:141)
	at org.springframework.retry.support.RetryTemplate.handleRetryExhausted(RetryTemplate.java:512)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:351)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:180)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:115)
	at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:153)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	
```

**Comment by dsyer** [comment]
Fixed in 9474abb
